✨ Add support for copying `contextvars` to thread workers (#2160)
python-trio/trio: docs/source/reference-core.rst

   request 0: Request received finished

For more information, read the
`contextvars docs <https://docs.python.org/3.7/library/contextvars.html>`__.

.. note::

    In advanced use cases you can also pass a custom `contextvars` context to
    ``nursery.start_soon(child, context=context)``.

>     That's an advanced use case, and you probably won't need that easily, but now you
>     know in case you need that in the future.

Reviewer: ```suggestion
    That's an advanced use case that you probably won't need, but now you
    know why that keyword argument exists!
```

.. literalinclude:: reference-core/from-thread-example.py

Threads and task-local storage
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When working with threads, you can use the same `contextvars` we discussed above,
because their values are preserved.

This is done by automatically copying the `contextvars` context when you use any of:

* `trio.to_thread.run_sync`
* `trio.from_thread.run`
* `trio.from_thread.run_sync`

That means that the values of the context variables are accessible even in worker
threads, or when sending a function to be run in the main/parent Trio thread using
`trio.from_thread.run` *from* one of these worker threads.

But it also means that as the context is not the same but a copy, if you `set` the
context variable value *inside* one of these functions that work in threads, the
new value will only be available in that context (that was copied). So, the new value
will be available for that function and other internal/children tasks, but the value
won't be available in the parent thread.

> If you need to modify values that would live in the context variables and you need to
> make those modifications from the children threads, you can instead set an object
> (e.g. a dictionary) in the context variable, in the top level/parent Trio thread. And
> then in the children, instead of setting the context variable, you can ``get`` the same
> object, and modify its values. That way you keep the same object in the context
> variable and only mutate it in children threads.
> 
> This way, you can modify the object content in children threads and still access the
> new content in the parent thread.

Reviewer: ```suggestion
If you need to modify values that would live in the context variables and you need to
make those modifications from the child threads, you can instead set a mutable object
(e.g. a dictionary) in the context variable of the top level/parent Trio thread.
Then in the children, instead of setting the context variable, you can ``get`` the same
object, and modify its values. That way you keep the same object in the context
variable and only mutate it in child threads.

This way, you can modify the object content in child threads and still access the
new content in the parent thread.
```
